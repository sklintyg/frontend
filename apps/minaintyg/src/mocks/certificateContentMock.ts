export const certificateContentMock = [
  {
    heading: 'Smittbärarpenning',
    body: `
      <h3>Avstängning enligt smittskyddslagen på grund av smitta</h3>
      <p>Ej angivet</p>
    `,
  },
  {
    heading: 'Grund för medicinskt underlag',
    body: `
      <h3>Intyget är baserat på</h3>
      <h4>Min undersökning av patienten</h4>
      <p>2023-02-07</p>
      <h4>Min telefonkontakt med patienten</h4>
      <p>Ej angivet</p>
      <h4>Journaluppgifter från den</h4>
      <p>Ej angivet</p>
      <h4>Ange vad annat är</h4>
      <p>Ej angivet</p>
    `,
  },
  {
    heading: 'Sysselsättning',
    body: `
      <h3>I relation till vilken sysselsättning bedömer du arbetsförmågan?</h3>
      <ul>
        <li>Nuvarande arbete</li>
      </ul>
      <h3>Ange yrke och arbetsuppgifter</h3>
      <p>Kontorsarbetare</p>

      <table>
        <thead>
          <tr>
            <th>Diagnoskod enligt ICD-10 SE</th>
            <th>Diagnos</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>S829</td>
            <td>Underbensfraktur med ospecificerad lokalisation</td>
          </tr>
        </tbody>
      </table>
    `,
  },
  {
    heading: 'Sjukdomens konsekvenser',
    body: `
      <h3>Funktionsnedsättning</h3>
      <h4>Beskriv undersökningsfynd, testresultat och observationer</h4>
      <p>Patienten har en underbensfraktur. Behandlas med helbensgrips och har därför nedsatt rörlighet i benet. Patienten
        får inte stödja på det skadade benet. Går långsamt och ostadigt med kryckor. Påtalar smärta vid undersökningen.
      </p>
      <h3>Aktivitetsbegränsning</h3>
      <h4>Beskriv vad patienten inte kan göra på grund av sin sjukdom. Ange vad uppgiften grundas
        på.</h4>
      <p>Patienten kan inte gå utan hjälpmedel. Klarar inte av att köra bil. Således mycket svårt att ta sig till arbetet.
        Dessutom problem med trappor på arbetsplatsen. Patienten kan även ha svårt att fokusera på sina arbetsuppgifter
        pga smärta.</p>
    `,
  },
  {
    heading: 'Medicinsk behandling',
    body: `
      <h3>Pågående medicinska behandlingar/åtgärder</h3>
      <h4>Ange vad syftet är och om möjligt tidplan samt ansvarig vårdenhet</h4>
      <p>Erhållit helbensgips, 0-belastning med kryckor. Smärtlindring.</p>
      <h3>Planerade medicinska behandlingar/åtgärder</h3>
      <h4>Ange vad syftet är och om möjligt tidplan samt ansvarig vårdenhet</h4>
      <p>Gips totalt 4 månader, initialt helbensgips. Rtg-kontroll efter 2, 4 och 6 veckor av frakturens läge. Om stabilt
        frakturläge vid 6 v kontroll byte till underbensgips. Rehabilitering efter avslutad gipsbehandling.</p>
    `,
  },
  {
    heading: 'Bedömning',
    body: `
      <h3>Min bedömning av patientens nedsättning av arbetsförmågan</h3>
      <table>
        <thead>
          <tr>
            <th>Nedsättningsgrad</th>
            <th>Från och med</th>
            <th>Till och med</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>100 procent</td>
            <td>2023-02-07</td>
            <td>2023-02-20</td>
          </tr>
        </tbody>
      </table>

      <h3>Patientens arbetsförmåga bedöms nedsatt längre tid än den som Socialstyrelsens
        försäkringsmedicinska beslutsstöd
        anger, därför att</h3>
      <p>Ej angivet</p>

      <h3>Kommer möjligheterna till återgång i arbete försämras om arbetstiden förläggs på annat
        sätt än att arbetstiden
        minskas lika mycket varje dag?</h3>
      <p>Ej angivet</p>

      <h3>Beskriv de medicinska skälen till att möjligheterna till återgång i arbete försämras</h3>
      <p>Ej angivet</p>

      <h3>Resor till och från arbetet med annat färdmedel än normalt kan göra det möjligt för
        patienten att återgå till
        arbetet under sjukskrivningsperioden.</h3>
      <p>Ej angivet</p>

      <h3>Prognos för arbetsförmåga utifrån aktuellt undersökningstillfälle</h3>
      <p>Patienten förväntas kunna återgå helt i nuvarande sysselsättning inom 1 månad.</p>
    `,
  },
  {
    heading: 'Åtgärder',
    body: `
      <h3>Här kan du ange åtgärder som du tror skulle göra det lättare för patienten att återgå i arbete</h3>
      <ul>
        <li>Arbetsanpassning</li>
      </ul>
      <h3>Här kan du ange fler åtgärder. Du kan också beskriva hur åtgärderna kan underlätta återgång i arbete.</h3>
      <p>Möjlighet till distansarbete i hemmet skulle underlätta mycket för patienten.</p>
    `,
  },
  {
    heading: 'Övriga upplysningar',
    body: `
      <h3>Övriga upplysningar</h3>
      <p>Ej angivet</p>
    `,
  },
  {
    heading: 'Kontakt',
    body: `
      <h3>Kontakt med Försäkringskassan</h3>
      <h4>Jag önskar att Försäkringskassan kontaktar mig</h4>
      <p>Ej angivet</p>
      <h4>Ange gärna varför du vill ha kontakt</h4>
      <p>Ej angivet</p>
    `,
  },
].map(({ heading, body }) => ({ heading, body: body.trim().replace(/\s+/g, ' ') }))
