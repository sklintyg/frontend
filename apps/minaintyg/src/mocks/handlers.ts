/* eslint-disable import/no-extraneous-dependencies */
import { fakeCertificateLabel, faker, fakerFromSchema } from '@frontend/fake'
import { rest } from 'msw'
import { certificateListItemSchema } from '../schema/certificateList.schema'
import { userSchema } from '../schema/user.schema'

export const handlers = [
  rest.get('/api/user', (_, res, ctx) =>
    res(
      ctx.status(200),
      ctx.json(
        fakerFromSchema(userSchema, {
          stringMap: {
            personName: faker.name.fullName,
          },
        })()
      )
    )
  ),
  rest.post('/api/certificates', (_, res, ctx) =>
    res(
      ctx.status(200),
      ctx.json({
        content: Array.from(
          { length: 5 },
          fakerFromSchema(certificateListItemSchema, {
            stringMap: {
              title: fakeCertificateLabel,
              certificateId: faker.datatype.uuid,
              timestamp: () => faker.date.between('2021-01-01T00:00:00.000Z', new Date().toISOString()).toISOString(),
            },
          })
        ),
      })
    )
  ),
]
