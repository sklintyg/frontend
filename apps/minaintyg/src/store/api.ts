import { getCookie } from '@frontend/utils'
import { isAnyOf, isPlainObject } from '@reduxjs/toolkit'
import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react'
import { AvailableFunction, Certificate, CertificateMetadata } from '../schema/certificate.schema'
import { CertificateFilterOptions } from '../schema/certificateListFilter.schema'
import { User } from '../schema/user.schema'
import { CertificateFilterState } from './slice/certificateFilter.slice'

export const api = createApi({
  reducerPath: 'api',
  baseQuery: fetchBaseQuery({
    baseUrl: '/api/',
    prepareHeaders: (headers) => {
      if (getCookie('XSRF-TOKEN')) {
        headers.set('X-XSRF-TOKEN', getCookie('XSRF-TOKEN') ?? '')
      }
      return headers
    },
  }),
  tagTypes: ['User', 'Certificate'],
  endpoints: (builder) => ({
    getUser: builder.query<User, void>({
      query: () => 'user',
      providesTags: ['User'],
    }),
    getCertificates: builder.query<{ content: CertificateMetadata[] }, CertificateFilterState>({
      query: (body) => ({
        url: 'certificate',
        method: 'POST',
        body: Object.fromEntries(Object.entries(body).map(([key, value]) => [key, [value]])),
      }),
      providesTags: (result) =>
        result
          ? [...result.content.map(({ id }) => ({ type: 'Certificate' as const, id })), { type: 'Certificate', id: 'LIST' }]
          : [{ type: 'Certificate', id: 'LIST' }],
    }),
    getCertificatesFilter: builder.query<CertificateFilterOptions, void>({
      query: () => 'filters',
    }),
    getCertificate: builder.query<Certificate & { availableFunctions: AvailableFunction[] }, { id: string }>({
      query: ({ id }) => `certificate/${id}`,
      transformResponse: ({ certificate, availableFunctions }: { certificate: Certificate; availableFunctions: AvailableFunction[] }) => ({
        ...certificate,
        availableFunctions,
      }),
      providesTags: (result) => (result ? [{ type: 'Certificate' as const, id: result.metadata.id }] : []),
    }),
    sendCertificate: builder.mutation<void, { id: string }>({
      query: ({ id }) => ({
        url: `certificate/${id}/send`,
        method: 'POST',
      }),
      invalidatesTags: (result, error, { id }) => (error ? [] : [{ type: 'Certificate', id }]),
    }),
  }),
})

export const { useGetCertificatesQuery, useGetCertificateQuery, useGetCertificatesFilterQuery, useSendCertificateMutation } = api

export const isFulfilledEndpoint = isAnyOf(...Object.values(api.endpoints).map((endpoint) => endpoint.matchFulfilled))
export const isRejectedEndpoint = isAnyOf(...Object.values(api.endpoints).map((endpoint) => endpoint.matchRejected))
export const hasResponse = (o: unknown): o is { response: Response } =>
  isPlainObject(o) && 'response' in o && o.response instanceof Response
