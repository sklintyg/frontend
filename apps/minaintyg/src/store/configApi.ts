import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react'
import { AppConfig } from '../schema/appConfig.schema'

export const configApi = createApi({
  reducerPath: 'api',
  baseQuery: fetchBaseQuery({
    baseUrl: '/appconfig/api/',
  }),
  tagTypes: ['User'],
  endpoints: (builder) => ({
    getAppConfig: builder.query<AppConfig, void>({
      query: () => 'app',
    }),
    getCertificateDescription: builder.query<CertificateDescription, void>({
      query: () => 'map',
    }),
  }),
})

export const { useGetAppConfigQuery, useGetCertificateDescriptionQuery } = configApi
