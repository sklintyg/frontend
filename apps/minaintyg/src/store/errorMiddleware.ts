import { randomUUID } from '@frontend/utils'
import { isRejectedWithValue, Middleware, MiddlewareAPI, SerializedError } from '@reduxjs/toolkit'
import { FetchBaseQueryError } from '@reduxjs/toolkit/dist/query'

export type ErrorObject = (FetchBaseQueryError & { id?: string }) | (SerializedError & { id?: string })

/**
 * Error handling middleware
 * https://redux-toolkit.js.org/rtk-query/usage/error-handling
 */
export const errorMiddleware: Middleware = (_: MiddlewareAPI) => (next) => (action) => {
  if (isRejectedWithValue(action)) {
    const id = randomUUID()
    // const { method, url } = action.meta.baseQueryMeta.request
    // const { status } = action.meta.baseQueryMeta.response

    // Append identifier to payload
    return next(Object.assign(action, { payload: { ...action.payload, id } }))
  }

  return next(action)
}
