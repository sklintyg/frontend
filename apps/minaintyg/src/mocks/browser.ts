/* eslint-disable import/no-extraneous-dependencies */
import { faker } from 'fake'
import { setupWorker } from 'msw'
import { handlers } from './handlers'

faker.setLocale('sv')

export const worker = setupWorker(...handlers)
