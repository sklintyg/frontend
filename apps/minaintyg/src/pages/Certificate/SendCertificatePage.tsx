import { IDSCard, IDSSpinner } from '@frontend/ids-react-ts'
import { skipToken } from '@reduxjs/toolkit/dist/query'
import { useParams } from 'react-router-dom'
import { PageHeading } from '../../components/PageHeading/PageHeading'
import { PageHeadingDescription } from '../../components/PageHeading/PageHeadingDescription'
import { useGetCertificateQuery } from '../../store/api'
import { CertificateInformation } from './components/CertificateInformation'
import { ReadCertificateError } from './components/ReadCertificateError'
import { SendCertificateActions } from './components/SendCertificateActions/SendCertificateActions'
import { AvailableFunctionsTypeEnum } from '../../schema/certificate.schema'
import {ReadMoreAboutAction} from "./components/ReadMoreAboutDialog/ReadMoreAboutAction";

export function SendCertificatePage() {
  const { id } = useParams()
  const { data: certificate, isLoading, error } = useGetCertificateQuery(id ? { id } : skipToken)
  const { recipient } = certificate?.metadata || {}
  const sendFunction = certificate?.availableFunctions.find(
    (availableFunction) => availableFunction.type === AvailableFunctionsTypeEnum.enum.SEND_CERTIFICATE
  )

  return (
    <>
      <PageHeading heading="Skicka intyg ">
        <PageHeadingDescription>{sendFunction?.body}</PageHeadingDescription>
        <ReadMoreAboutAction />
      </PageHeading>
      {isLoading && <IDSSpinner data-testid="spinner" />}
      {error && <ReadCertificateError id={id} error={error} />}
      {certificate && recipient && sendFunction && (
        <>
          <div>
            <h2 className="ids-heading-2 mb-5">Intyg som ska skickas</h2>
            <IDSCard className="mb-5">
              <h2 className="ids-heading-3">{certificate.metadata.type.name}</h2>
              <CertificateInformation {...certificate.metadata} />
            </IDSCard>
          </div>
          <div className="mb-5">
            <h2 className="ids-heading-2 mb-5">Mottagare</h2>
            <IDSCard>{recipient.name}</IDSCard>
          </div>
          {id && <SendCertificateActions id={id} recipient={recipient} />}
        </>
      )}
    </>
  )
}
