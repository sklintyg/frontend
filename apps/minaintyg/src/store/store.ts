import { configureStore } from '@reduxjs/toolkit'
import { api } from './api'
import { errorMiddleware } from './middleware/error.middleware'
import { reducer } from './reducer'
import { idpMiddleware } from './slice/idpMiddleware'

export const store = configureStore({
  reducer,
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat([api.middleware, idpMiddleware, errorMiddleware]),
})

export type AppDispatch = typeof store.dispatch
