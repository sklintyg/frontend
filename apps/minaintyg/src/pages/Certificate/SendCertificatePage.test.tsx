import { faker, fakerFromSchema } from '@frontend/fake'
import { render, screen, waitFor } from '@testing-library/react'
import { rest } from 'msw'
import { Provider } from 'react-redux'
import { Route, RouterProvider, createMemoryRouter, createRoutesFromChildren } from 'react-router-dom'
import { expect, it } from 'vitest'
import { SendCertificatePage } from './SendCertificatePage'
import { server } from '../../mocks/server'
import {
  AvailableFunctionsTypeEnum,
  certificateMetadataSchema,
  certificateRecipientSchema,
  certificateSchema,
} from '../../schema/certificate.schema'
import { store } from '../../store/store'

function renderComponent() {
  return render(
    <Provider store={store}>
      <RouterProvider
        router={createMemoryRouter(createRoutesFromChildren([<Route key="root" path="/:id" element={<SendCertificatePage />} />]), {
          initialEntries: ['/12345'],
        })}
      />
    </Provider>
  )
}

it('Should render as expected', async () => {
  faker.seed(412341)
  server.use(
    rest.get('/api/certificate/:id', (_, res, ctx) =>
      res(
        ctx.json({
          certificate: fakerFromSchema(certificateSchema)({
            metadata: fakerFromSchema(certificateMetadataSchema)({
              id: '16eb261e-545c-41db-8a6c-59133ead80a4',
              events: [],
              issued: '2023-10-10T13:37:00.000Z',
              issuer: {
                name: 'Jenny Ström',
              },
              summary: {
                label: 'Gäller intygsperiod',
                value: '2021-04-22 - 2021-07-19',
              },
              recipient: fakerFromSchema(certificateRecipientSchema)({
                name: 'Transportstyrelsen',
                sent: null,
              }),
            }),
          }),
          availableFunctions: [
            {
              type: AvailableFunctionsTypeEnum.enum.SEND_CERTIFICATE,
              body: 'Skicka intyget body',
              enabled: true,
            },
          ],
        })
      )
    )
  )
  const { container } = renderComponent()
  expect(await screen.findByTestId('spinner')).toBeInTheDocument()
  await waitFor(() => expect(screen.queryByText('spinner')).not.toBeInTheDocument())
  expect(container).toMatchSnapshot()
})

it('Should render error message when unable to load certificate', async () => {
  server.use(rest.get('/api/certificate/:id', (_, res, ctx) => res(ctx.status(500))))
  renderComponent()
  await waitFor(() => expect(screen.queryByTestId('spinner')).not.toBeInTheDocument())
  expect(screen.getAllByRole('alert')).toMatchSnapshot()
})
