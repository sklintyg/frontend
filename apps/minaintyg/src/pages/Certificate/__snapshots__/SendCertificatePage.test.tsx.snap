// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Should render as expected 1`] = `
<div>
  <header
    class="ids-content mb-7"
  >
    <h1
      class="ids-heading-1"
    >
      Skicka intyg 
    </h1>
    <div
      class="ids-preamble mb-2 text-lg md:text-[1.375rem]/[2.25rem]"
    >
      Från den här sidan kan du välja att skicka ditt intyg digitalt till mottagaren. Endast mottagare som kan ta emot digitala intyg visas nedan.
    </div>
  </header>
  <div>
    <h2
      class="ids-heading-2 mb-5"
    >
      Intyg som ska skickas
    </h2>
    <ids-card
      class="mb-5"
    >
      <h2
        class="ids-heading-3"
      >
        Dr. Robert Svensson
      </h2>
      <div
        class="flex flex-col gap-5 md:flex-row md:gap-10"
      >
        <div>
          <h3
            class="ids-heading-4 mb-0"
          >
            Gäller intygsperiod
          </h3>
          2021-04-22 - 2021-07-19
        </div>
        <div>
          <h3
            class="ids-heading-4 mb-0"
          >
            Intyg utfärdat
          </h3>
          2023-10-10
        </div>
        <div>
          <h3
            class="ids-heading-4 mb-0"
          >
            Skrivet av
          </h3>
          Jenny Ström
          , 
          Tommy Claesson
        </div>
      </div>
    </ids-card>
  </div>
  <div
    class="mb-5"
  >
    <h2
      class="ids-heading-2 mb-5"
    >
      Mottagare
    </h2>
    <ids-card>
      Transportstyrelsen
    </ids-card>
  </div>
  <div
    class="flex flex-col gap-5 py-5 sm:flex-row"
  >
    <ids-button
      role="button"
      sblock=""
      secondary=""
      size="m"
      tabindex="0"
      type=""
    >
      Tillbaka till intyget
    </ids-button>
    <ids-button
      role="button"
      sblock=""
      size="m"
      tabindex="0"
      type=""
    >
      Skicka
    </ids-button>
  </div>
</div>
`;
